pkg/process/util/containers.go

```diff
@@ -6,104 +6,280 @@
-		detectors = []*collectors.Detector{collectors.NewDetector("")}
-	}
-
-	result := []*containers.Container{}
-	for _, detector := range detectors {
-		l, name, err := detector.GetPreferred()
+func (p *ContainerProvider) GetContainers(cacheValidity time.Duration, previousContainers map[string]*ContainerRateMetrics, previousTime, currentTime time.Time) ([]*model.Container, map[string]*ContainerRateMetrics, map[int]string, error) {
+	containersMetadata, err := p.metadataStore.ListContainers()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	hostCPUCount := float64(system.HostCPUCount())
+	processContainers := make([]*model.Container, 0)
+	rateStats := make(map[string]*ContainerRateMetrics)
+	pidToCid := make(map[int]string)
+	for _, container := range containersMetadata {
+		if !container.State.Running {
+			continue
+		}
+
+		if p.filter != nil && p.filter.IsExcluded(container.Name, container.Image.Name, container.Labels["io.kubernetes.pod.namespace"]) {
+			continue
+		}
+
+		if container.Runtime == workloadmeta.ContainerRuntimeGarden && len(container.CollectorTags) == 0 {
+			log.Debugf("No tags found for garden container: %s, skipping", container.ID)
+			continue
+		}
+
+		entityID := containers.BuildTaggerEntityName(container.ID)
+		tags, err := tagger.Tag(entityID, collectors.HighCardinality)
 		if err != nil {
-			return nil, err
+			log.Warnf("Could not collect tags for container %q, err: %w", container.ID[:12], err)
+		}
+		tags = append(tags, container.CollectorTags...)
+
+		outPreviousStats := NullContainerRates
+		// Name and Image fields exist but are never filled
+		processContainer := &model.Container{
+			Type:      string(container.Runtime),
+			Id:        container.ID,
+			Started:   container.State.StartedAt.Unix(),
+			Created:   container.State.CreatedAt.Unix(),
+			Tags:      tags,
+			State:     convertContainerStatus(container.State.Status),
+			Health:    convertHealthStatus(container.State.Health),
+			Addresses: computeContainerAddrs(container),
 		}
-		// Get containers from cache and update metrics
-		cacheKey := cache.BuildAgentKey("containers", name)
-		cached, hit := cache.Cache.Get(cacheKey)
-		if hit {
-			containers, ok := cached.([]*containers.Container)
-			if ok {
-				err := l.UpdateMetrics(containers)
-				if err != nil {
-					log.Debugf("Cannot update container metrics via %s: %s", name, err)
-					continue
-				}
-				log.Tracef("Got %d containers from cache", len(containers))
-				result = append(result, containers...)
-				continue
+
+		// Gathering container & network statistics
+		previousContainerRates := previousContainers[container.ID]
+		if previousContainerRates == nil {
+			previousContainerRates = &NullContainerRates
+		}
+
+		collector := p.metricsProvider.GetCollector(string(container.Runtime))
+		if collector == nil {
+			log.Infof("No metrics collector available for runtime: %s, skipping container: %s", container.Runtime, container.ID)
```

Comment:
This also may result in log spam.

Reply:
Not sure I agree there, it would mean we have implemented a collector in `workloadmeta` without implementing a metrics collector, which should never happen.
I think it's worth logging it as INFO even though it will indeed log at every check run.